include/pybind11/detail/internals.h

```diff
@@ -419,11 +419,33 @@ PYBIND11_NOINLINE internals &get_internals() {
     } gil;
     error_scope err_scope;
 
-    PYBIND11_STR_TYPE id(PYBIND11_INTERNALS_ID);
-    auto builtins = handle(PyEval_GetBuiltins());
-    if (builtins.contains(id) && isinstance<capsule>(builtins[id])) {
-        internals_pp = static_cast<internals **>(capsule(builtins[id]));
+    constexpr const char *id_cstr = PYBIND11_INTERNALS_ID;
+    str id(id_cstr);
 
+    dict state_dict;
+#if PY_VERSION_HEX < 0x03080000 || defined(PYPY_VERSION)
```

Comment:
@rwgk Should we ping PyPy to see if we can this fixed in the next release?

Reply:
Not sure. Might be a bit early right now. After this PR is merged might be a better timing for asking. Which brings me to another thought:

It would actually be great to merge this PR asap, but with `PY_VERSION_HEX < 0x030C0000`, so that we get Python 3.12 on the right track even before day 1, never will have to worry about related backwards compatibility questions for anything >= 3.12, and pre-release testing with 3.12 will never run into issues around using `builtins`.

Reply:
Does pybind11 actually compile with Python 3.12? My guess would be some internal details will need to change just like with 3.11.

Reply:
Edit: it compiles with `cpython/main`, though there are apparently some new failures in the test suite.
```
RROR tests/test_custom_type_setup.py::test_self_cycle - pytest.PytestUnraisableExceptionWarning: Exception ignor...
ERROR tests/test_custom_type_setup.py::test_indirect_cycle - pytest.PytestUnraisableExceptionWarning: Exception i...
FAILED tests/test_call_policies.py::test_alive_gc - pytest.PytestUnraisableExceptionWarning: Exception ignored in...
FAILED tests/test_methods_and_attributes.py::test_dynamic_attributes - pytest.PytestUnraisableExceptionWarning: E...
FAILED tests/test_methods_and_attributes.py::test_cyclic_gc - pytest.PytestUnraisableExceptionWarning: Exception ...
FAILED tests/test_multiple_inheritance.py::test_mi_dynamic_attributes - pytest.PytestUnraisableExceptionWarning: ...
FAILED tests/test_pickling.py::test_roundtrip_with_dict[PickleableWithDict] - pytest.PytestUnraisableExceptionWar...
FAILED tests/test_pickling.py::test_roundtrip_with_dict[PickleableWithDictNew] - pytest.PytestUnraisableException...
```

Reply:
I would say that wide-scale deployment of 3.12 is likely so far off that an ABI bump in pybind11 is perhaps the closer thing to aim for with regards to this PR.